src/west/commands/project.py

```diff
@@ -865,12 +883,12 @@ def _fetch(project):
     update_cmd = 'update-ref ' + QUAL_MANIFEST_REV + ' '
     if _maybe_sha(project.revision):
         # Don't fetch a SHA directly, as server may restrict from doing so.
-        fetch_cmd += 'refs/heads/*:refs/west/*'
+        fetch_cmd += 'refs/heads/*:' + WEST_REF + '*'
         update_cmd += '{revision}'
     else:
         # The revision is definitely not a SHA, so it's safe to fetch directly.
         # This avoids fetching unnecessary ref space from the remote.
-        fetch_cmd += '{revision}'
+        fetch_cmd += '{revision}:' + WEST_REF + '{revision}'
```

Comment:
Did we agree to use a reflog on refs/heads/manifest-rev to keep track of old versions of the `revision` field? That's what I thought we agreed to.

If so, why can't we just update manifest-rev directly in this case? That would avoid the need to update the west refspace entirely. We could still keep the code for SHAs as-is, but then also clean up after ourselves after we're done:

```
diff --git a/src/west/commands/project.py b/src/west/commands/project.py
index ca4361f..01c5e4d 100644
--- a/src/west/commands/project.py
+++ b/src/west/commands/project.py
@@ -852,6 +852,9 @@ def _fetch(project):
     msg = "{name}: fetching, need revision {revision}"
     if project.clone_depth:
         msg += " with --depth {clone_depth}"
+
+    log.small_banner(project.format(msg))
+
     # -f is needed to avoid errors in case multiple remotes are present,
     # at least one of which contains refs that can't be fast-forwarded to our
     # local ref space.
@@ -860,20 +863,20 @@ def _fetch(project):
     fetch_cmd = ('fetch -f --tags ' +
                  ('--depth={clone_depth} ' if project.clone_depth else ' ') +
                  '-- {url} ')
-    update_cmd = 'update-ref ' + QUAL_MANIFEST_REV + ' '
     if _maybe_sha(project.revision):
         # Don't fetch a SHA directly, as server may restrict from doing so.
         fetch_cmd += 'refs/heads/*:refs/west/*'
-        update_cmd += '{revision}'
+        update_cmd = 'update-ref ' + QUAL_MANIFEST_REV + ' {revision}'
+        project.git(fetch_cmd)
+        project.git(update_cmd)
+        _clean_west_refspace(project)
     else:
-        # The revision is definitely not a SHA, so it's safe to fetch directly.
-        # This avoids fetching unnecessary ref space from the remote.
-        fetch_cmd += '{revision}'
-        update_cmd += 'FETCH_HEAD^{{commit}}'
-
-    log.small_banner(project.format(msg))
-    project.git(fetch_cmd)
-    project.git(update_cmd)
+        # The revision is definitely not a SHA, so it's safe to fetch
+        # directly into manifest-rev. This avoids fetching unnecessary
+        # refs from the remote, which is faster and saves bandwidth.
+        fetch_cmd += '{revision}:' + QUAL_MANIFEST_REV
+        update_cmd = None
+        project.git(fetch_cmd)
 
     if not _head_ok(project):
         # If nothing it checked out (which would usually only happen just after
```

Fetching directly into manifest-rev leaves a nice clean entry in my reflog:

```
$ git reflog -n1 manifest-rev
132130d17a (manifest-rev) manifest-rev@{0}: fetch https://github.com/zephyrproject-rtos/zephyr +master:refs/heads/manifest-rev: forced-update
```

Bonus, doing it that way means we can just delete *all* the `refs/west/*` using a bulk git operation after updating a "maybe-SHA" instead of looping over them individually, like discussed in this SO: https://stackoverflow.com/a/46229416

That would be faster on Windows, where process creation is expensive.

Reply:
(Above patch is untested and contains a couple of bad lines, like the `update_cmd = None`, but hopefully makes my idea clear.)

Reply:
> Did we agree to use a reflog on refs/heads/manifest-rev to keep track of old versions of the revision field? That's what I thought we agreed to.
> 
> If so, why can't we just update manifest-rev directly in this case? 

not sure we actually discussed the reflog part, but that dates back a long time ;)
But you have a valid point that we can actually fetch directly into `manifest-rev` without having to go through `FETCH_HEAD`. Not sure why no one considered that in the first place.

But this doesn't address my main point with this PR, which is actually to be able to verify what revision is currently in use.
With fetching directly `manifest-rev` the user (and other extension commands) can no longer determine if the current `manifest-rev` revision is identical to current manifest file, cause only the SHA is available.

So if I do `git log`, `west list zephyr`, or `west status zephyr`  I might see:
```
$ git log --oneline -n1
b687d5d022 (HEAD, manifest-rev) [nrf noup] settings: Fix NVS initialization with PM enabled

$ west status zephyr
=== status of zephyr (zephyr):
HEAD detached at b687d5d022
nothing to commit, working tree clean

$ west list zephyr
zephyr     zephyr     pull/240/head     https://github.com/NordicPlayground/fw-nrfconnect-zephyr
```

but I have no idea whether `b687d5d022` is actually identical to `pull/240/head` or not.
Now, with the proposal in this PR, of keeping only the last updated refs under `refs/west` when it's not a SHA, it provides this:
```
$ git log --oneline -n1
a6a98b7854 (HEAD, refs/west/pull/240/head, manifest-rev) [nrf noup] partition-manager: Fix s1_variant dependency
```
i'm looking at `pull/240/head`, Good to know :)

```
$ git rev-parse refs/west/pull/240/head
a6a98b7854c8ff4e1af7bf22d991379535284521
```
which can then compare to either `HEAD` or current `manifest-rev`, which also means we could do:
```
$ west list zephyr
zephyr  zephyr   pull/240/head (not checked out)    https://github.com/NordicPlayground/fw-nrfconnect-zephyr
```
to indicate to the user that the manifest file has changed, but that he is NOT looking at revision `pull/240/head`, or we could add this info to `west status` instead.

Just as in: https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1437
that you now have closed, you would be able to determine if current `manifest revision exists in refs/west/{revision}` and thus avoid the additional fetch.

If we only have `manifest-rev` there is no relation back to the manifest file.
With both manifest-rev, `west.yml` and a single `refs/west/<revision>` when not a SHA, we can the do:
- is `manifest-rev` == `west.yml::revision` and available in `refs/west` space

Reply:
> not sure we actually discussed the reflog part, but that dates back a long time ;)

OK, let me try to summarize the discussion from what I understood, because I think that we have misunderstood each other again.

I wrote a concern about garbage collection of items in refs/west:

https://github.com/zephyrproject-rtos/west/pull/331#discussion_r350387809

>> What are the concern(s)?
>>
>> I guess, user will never need to worry or care about our additional refs/west space.

> We're creating an unbounded amount of un-garbage-collectable refs here, including putting tags outside of refs/tags.

You responded:

https://github.com/zephyrproject-rtos/west/pull/331#discussion_r351196763

> We should look into cleaning up this ref-space.

Marc responded with a comment about the reflog:

https://github.com/zephyrproject-rtos/west/pull/331#issuecomment-558382858

> Connecting all the dots above, I just remembered that git already has a built-in feature to cache objects and automatically expire and garbage collect them after some configurable time: it's the reflog. 

I responded to him:

https://github.com/zephyrproject-rtos/west/pull/331#issuecomment-558635485

> Good point!
>
> Since manifest-rev is reflogged already, do we need this then?

You responded:

https://github.com/zephyrproject-rtos/west/pull/331#issuecomment-559024596

> @marc-hb Good point on updating the reflog, and thus cleaning the refs/west will also help te ensure the SHAs are available for a certain time, while not have all those refs/west references popping up in git log.
> As I reply in #331 (comment) , there needs to be done some cleanup in refs/west, and this appears to be a nice way of doing that.

I assumed by your "this appears to be a nice way of doing that" in response to my question "do we need this", (and by "this", I mean this PR, #331), that you agreed we did not need to keep storing things in refs/west, and we can just use the reflog on the manifest-rev branch.

OK, so that's it so far from my side.

-------------------------------------------

> which can then compare to either HEAD or current manifest-rev, which also means we could do:

```
$ west list zephyr
zephyr  zephyr   pull/240/head (not checked out)    https://github.com/Nor
```

Now given this last example I think I finally see what you want to do with this, and what you really meant in the original pull description when you wrote:

> To allow later usage of the fetched refs space then git fetch will
> always keep a local ref space even fetching non-SHA refs spaces, such
> as branches and PRs.

I really am not sure this is within the scope of west. We certainly originally tried hard to avoid creating any refs at all other than `manifest-rev`, so I think we need to decide what the "real feature" is before we continue to create refs that will show up in people's git log. I personally do not like the refs/west showing up in my git logs, at least.

As you've stated, we started using `refs/west/` as a workaround to fetch SHAs, not because we wanted to track refnames from the remote there.

So can we talk about this in terms of what you ultimately want to do?

Reply:
> > not sure we actually discussed the reflog part, but that dates back a long time ;)
>
> OK, let me try to summarize the discussion from what I understood, because I think that we have misunderstood each other again.
> 
> I wrote a concern about garbage collection of items in refs/west:

Ahh, now I get it. I thought you were referring back to when we introduced `manifest-rev`, and not the fact pointed out by @marc-hb.

> Now given this last example I think I finally see what you want to do with this, and what you really meant in the original pull description when you wrote:

Yes, this was my original intention with this PR, and then it got sidetracked into cleaning up `refs/west` for all the revisions that was purely kept under `refs/west` because of SHA fetching.
(And a good sidetracking I believe)

So original intention was to keep manifest revision available under `refs/west/<revision>` when it is not a SHA. (and remove all other revisions under `refs/west`).
I found such functionality useful for minimizing need for additional fetching in NordicPlayground/fw-nrfconnect-nrf#1437  and considered it to be potentially useful in other places.
But i'm of course open to hear all opinions on whether this is a valid use-case.


For the other revision under `refs/west` I fully agree that we should not keep those available under `refs/west` just because we need to fetch SHAs.

So that's why I included this cleanup into this PR so that `refs/west` is always cleaned up when pointing to a SHA :-)
Although it was actually a side-track to the original PR.

Maybe this PR should now be split into 2:
1) Cleaning of `refs/west` ref space when fetching SHAs - We definately want this
2) Keep current `refs/west/<revision>` for non-SHA - As this is controversial

Reply:
> Maybe this PR should now be split into 2:
> 
>     1. Cleaning of `refs/west` ref space when fetching SHAs - We definately want this

+1 to this

>     2. Keep current `refs/west/<revision>` for non-SHA - As this is controversial

And this. I don't mind keeping `refs/west/<revision>` if it helps use cases that matter to people. I closed https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1437 unmerged because I think all I need is `west fetch` as you suggested, not `refs/west/xyz`.

Reply:
> 
>
>   2. Keep current `refs/west/<revision>` for non-SHA - As this is controversial
>
>And this. I don't mind keeping refs/west/<revision> if it helps use cases that matter to people. I closed NordicPlayground/fw-nrfconnect-nrf#1437 unmerged because I think all I need is west fetch as you suggested, not refs/west/xyz.

Fair enough.
This was suggested so that you could avoid a `west fetch` if `revision: pull/<no>/head` was already available.
Which would allow us to stick to current behavior where only `west update` actually touches the network ;)

Reply:
Hmm, we could do `west update --fetch-only` or so if that's a concern.

Reply:
> Hmm, we could do `west update --fetch-only` or so if that's a concern.

It's not really a concern for me, as such.

If I remember correctly, you have previously mentioned that today only `west update` touches the network and make changes to projects.

Also you have in general have preferred to minimize fetching.

And I do believe both principles above are good to have in mind when working on `west`.

That was the reason for my comment.

Sorry if I have misunderstood, and I'm referring your intentions wrongly :flushed:

Reply:
You are right that I prefer to minimize fetching. It's true that right now, west update is the only thing that touches the network (other than west init, of course), but I think it would be OK to add more if there is a good reason. I do want to minimize the number of built-in commands, but it seems like a `west fetch` would be reasonable in this case.